torch/optim/optimizer.py

```diff
@@ -34,6 +35,18 @@ def __init__(self, params, defaults):
         torch._C._log_api_usage_once("python.optimizer")
         self.defaults = defaults
 
+        def _step_with_profile(func, class_name):
+            """Wrap self.step under profiler. In order to avoid of instrumenting each sub-class."""
+            @functools.wraps(func)
+            def wrapper(*args, **kwargs):
+                name = str.format("{}#{}.{}", "Optimizer.step", class_name, "step")
+                with torch.autograd.profiler.record_function(name):
+                    return func(*args, **kwargs)
+            return wrapper
+        # Using this way could elegantly wrap self.step as a method not function,
+        # which don't influence re-wrapped by _LRScheduler and deepcopy.
+        self.__class__.step = _step_with_profile(self.__class__.step, self.__class__.__name__)
```

Comment:
wait a sec - you're editing a class on every instance creation. Won't it create many many nested wrappers?

E.g. if someone creates 1000 optimizers (or for some reason creates an optimizer in a loop), I think you're going to get 1000 of nested wrappers :( 

You probably want some detection to avoid repeated patching. Maybe try `.step.qualname`? (not sure whether it works well or not

Reply:
Thanks a lot! Fixed by checking a flag, now only the first object trigger it.